packages/common/solana/utils.ts

```diff
@@ -125,72 +142,125 @@ export async function signTransaction(invoker: Keypair | SignerWalletAdapter, tx
  * @param connection - Solana client connection
  * @param invoker - Keypair used as signer
  * @param tx - Transaction instance
- * @param hash - blockhash information, the same hash should be used in the Transaction
+ * @param {ConfirmationParams} confirmationParams - Confirmation Params that will be used for execution
  * @returns Transaction signature
  */
 export async function signAndExecuteTransaction(
   connection: Connection,
   invoker: Keypair | SignerWalletAdapter,
-  tx: Transaction,
-  hash: BlockhashWithExpiryBlockHeight,
+  tx: Transaction | VersionedTransaction,
+  confirmationParams: ConfirmationParams,
 ): Promise<string> {
   const signedTx = await signTransaction(invoker, tx);
 
-  return executeTransaction(connection, signedTx, hash);
+  return executeTransaction(connection, signedTx, confirmationParams);
 }
 
 /**
  * Sends and confirms Transaction
- * Confirmation strategy is not 100% reliable here as in times of congestion there can be a case that tx is executed,
- * but is not in `commitment` state and so it's not considered executed by the `sendAndConfirmRawTransaction` method,
- * and it raises an expiry error even though transaction may be executed soon.
- * - so we add additional 50 blocks for checks to account for such issues;
- * - also, we check for SignatureStatus one last time as it could be that websocket was slow to respond.
+ * Uses custom confirmation logic that:
+ * - simulates tx before sending separately
+ * - sends transaction without preFlight checks but with some valuable flags https://twitter.com/jordaaash/status/1774892862049800524?s=46&t=bhZ10V0r7IX5Lk5kKzxfGw
+ * - rebroadcasts a tx every 500 ms
+ * - after broadcasting check whether tx has executed once
+ * - catch errors for every actionable item, throw only the ones that signal that tx has failed
+ * - otherwise there is a chance of marking a landed tx as failed if it was broadcasted at least once
  * @param connection - Solana client connection
  * @param tx - Transaction instance
  * @param hash - blockhash information, the same hash should be used in the Transaction
+ * @param context - context at which blockhash has been retrieve
+ * @param commitment - optional commitment that will be used for simulation and confirmation
  * @returns Transaction signature
  */
 export async function executeTransaction(
   connection: Connection,
-  tx: Transaction,
-  hash: BlockhashWithExpiryBlockHeight,
+  tx: Transaction | VersionedTransaction,
+  { hash, context, commitment }: ConfirmationParams,
 ): Promise<string> {
-  const rawTx = tx.serialize();
+  if (!hash.lastValidBlockHeight || tx.signatures.length === 0 || !hash.blockhash) {
+    throw Error("Error with transaction parameters.");
+  }
+
+  for (let i = 0; i < 3; i++) {
+    let res: RpcResponseAndContext<SimulatedTransactionResponse>;
+    if (isTransactionVersioned(tx)) {
+      res = await connection.simulateTransaction(tx);
```

Comment:
I had the same thought, but looks like method overloading in TS works differently and it's incapable of handling this correctly. Therefore, this will produce an error because of polymorphic input. I guess there is some way of casting `tx` correctly, but it won't make this place easier to read 😔

Reply:
What's more likely btw, solana.web3 produces incomplete d.ts file for this function